tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -68,258 +94,410 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
         None => FunctionNameMatch::Anything,
     };
 
-    // Configure a thread pool with a larger stack size to prevent overflowing stack in large programs.
-    // Default is 2MB.
-    let pool = rayon::ThreadPoolBuilder::new().stack_size(4 * 1024 * 1024).build().unwrap();
-    let test_reports: Vec<Vec<(String, TestStatus)>> = pool.install(|| {
-        workspace
-            .into_iter()
-            .par_bridge()
-            .map(|package| {
-                run_tests::<Bn254BlackBoxSolver>(
-                    &workspace_file_manager,
-                    &parsed_files,
-                    package,
-                    pattern,
-                    args.show_output,
-                    args.oracle_resolver.as_deref(),
-                    Some(workspace.root_dir.clone()),
-                    Some(package.name.to_string()),
-                    &args.compile_options,
-                )
-            })
-            .collect::<Result<_, _>>()
-    })?;
-    let test_report: Vec<(String, TestStatus)> = test_reports.into_iter().flatten().collect();
-
-    if test_report.is_empty() {
-        match &pattern {
-            FunctionNameMatch::Exact(pattern) => {
-                return Err(CliError::Generic(
-                    format!("Found 0 tests matching input '{pattern}'.",),
-                ))
+    let num_threads = args.test_threads.unwrap_or_else(rayon::current_num_threads);
+
+    let runner = TestRunner {
+        file_manager: &file_manager,
+        parsed_files: &parsed_files,
+        workspace,
+        args: &args,
+        pattern,
+        num_threads,
+    };
+    runner.run()
+}
+
+struct TestRunner<'a> {
+    file_manager: &'a FileManager,
+    parsed_files: &'a ParsedFiles,
+    workspace: Workspace,
+    args: &'a TestCommand,
+    pattern: FunctionNameMatch<'a>,
+    num_threads: usize,
+}
+
+impl<'a> TestRunner<'a> {
+    fn run(&self) -> Result<(), CliError> {
+        // First compile all packages and collect their tests
+        let package_test_results = self.collect_packages_tests();
```

Comment:
When I read `package_test_results` I think _results_ as in the results of test execution, a collection of `TestResult` type above. I see it's actually a `HashMap<String, Result<Vec<Test<'a>>, CliError>>`. 

I see why you return `Result`, so we inspect compilation failures on a package-by-package basis, I reckon just `package_tests` would be a more intuitive name, especially because in iteration you use `(package_name, package_tests)`, not e.g. `(package_name, package_tests_result)`.